R/data_partition.R

```diff
@@ -47,35 +64,47 @@ data_partition <- function(data,
     set.seed(seed)
   }
 
-  training <- data.frame()
-  test <- data.frame()
+  if(sum(prob) > 1) {
+    stop("`prob` cannot be higher than 1.")
+  }
+
+  # Initialize out list
+  out <- list()
+  data_copy <- data
+  partitions <- rep(0, nrow(data))
 
-  if (!is.null(group)) {
-    for (i in split(data, data[group])) {
-      out <- .data_partition(i, training_proportion)
-      training <- rbind(training, i[out$training, ])
-      test <- rbind(test, i[out$test, ])
-    }
-  } else {
-    out <- .data_partition(data, training_proportion)
-    training <- rbind(training, data[out$training, ])
-    test <- rbind(test, data[out$test, ])
+  # Assign to same group if no groups
+  if(is.null(group)) {
+    data_copy$.temp <- "TEMP"
+    group <- ".temp"
   }
 
-  list(
-    training = training,
-    test = test
-  )
-}
+  # Store row index
+  data_copy$.rowid <- 1:nrow(data_copy)
 
+  # Iterate through probabilities (in case there is more than 1)
+  for(i in 1:length(prob)) {
+    out[i] <- c(NA) # Initialize empty value (will be dropped later)
+    # Split into groups, and for each group, get the desired proportion
+    for (dat in split(data_copy, data_copy[group])) {
+      out[[i]] <- c(out[[i]], dat$.rowid[sample(1:nrow(dat), size = prob[i] * nrow(dat))])
+    }
+    # Drop NA
+    out[[i]] <- as.numeric(na.omit(out[[i]]))
+    # Store partition result
+    partitions[out[[i]]] <- i
+    # Delete used rows from pool
+    data_copy <- data_copy[-out[[i]], ]
```

Comment:
In the current implementations, it basically loops through the possible values of `prob`, however, it removes everytime the selected rows from the original pool of data (to be without replacement (though it would be nice to have a "with replacement" option too). But because it does that, the size of the remaining pool changes, and the proportions differ. 

So when querying say `prob=c(0.3, 0.3)`, the first time it takes 30% of the dataframe as expected, but the next iteration it takes 30% from 100-30% of the data, resulting in less rows. 

I spent some time thinking how to fix this by adjusting the probabilities, but it just seems to convoluted, and I wonder if there's just a simpler way of partitioning data (with the possibility of stratification)...

@strengejacke @mattansb @bwiernik  fresh pair of eyes is welcome :)

Reply:
What if you convert the probs into a number of rows to keep before running the loop? Something like:
```r
rows_per_prob <- prob * nrow(data_copy)
  
  for(i in 1:length(rows_per_prob)) {
    out[i] <- c(NA) # Initialize empty value (will be dropped later)
    # Split into groups, and for each group, get the desired proportion
    for (dat in split(data_copy, data_copy[group])) {
      out[[i]] <- c(out[[i]], dat$.rowid[sample(1:nrow(dat), size = rows_per_prob[i])])
    }
...
``` 
``` r
data <- iris
data$Smell <- rep(c("Strong", "Strong", "Light"), 50)

x <- data_partition(data, prob=c(0.3, 0.3))

## Each subsample has 30% of the rows
nrow(x[[1]])
#> [1] 45
nrow(x[[2]])
#> [1] 45

## No rows are in both subsamples
length(which(x[[1]] %in% x[[2]]))
#> [1] 0

### Second example
y <- data_partition(data, prob=c(0.25, 0.4, 0.1))
nrow(y[[1]])
#> [1] 37
nrow(y[[2]])
#> [1] 54
nrow(y[[3]])
#> [1] 15
length(which(y[[1]] %in% y[[2]]))
#> [1] 0
length(which(y[[1]] %in% y[[3]]))
#> [1] 0
length(which(y[[2]] %in% y[[3]]))
#> [1] 0
```

<sup>Created on 2022-06-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Not sure it would work with groups though...

Reply:
A more efficient option might be to sample row indices as one step, then subset using those indices as a second set. 

Something like:

```r
n_obs <- nrow(dat)
counts <- sapply(probs, \(p) p * n_obs)
seq_n <- seq_len(n_obs)

indices <- vector("list", length(probs))

indices[1] <- sample(seq_n, counts[1])
for (i in seq_along(indices)[-1]) {
  indices[i] <- sample(setdiff(seq_n, unlist(indices[seq_len(i - 1)])), counts[i])
}

Reply:
Isn't it something like

``` r
# multiple probs
prob <- c(.3, .5)
d <- iris
lapply(prob, function(i) {
  sort(sample(1:nrow(d), round(nrow(d) * i), FALSE))
})
#> [[1]]
#>  [1]   4   7  13  14  15  22  26  32  33  36  38  41  44  46  53  54  55  58  66
#> [20]  70  71  73  83  87  89  92  94  96  97  98 100 112 119 123 124 129 130 133
#> [39] 134 136 138 139 143 145 148
#> 
#> [[2]]
#>  [1]   1   2   3   4   5   7   8  12  13  17  18  19  20  21  23  24  25  27  29
#> [20]  33  35  36  37  38  42  46  48  49  51  52  54  57  58  61  62  63  65  66
#> [39]  67  70  71  72  73  74  78  82  83  88  89  91  92  98 104 105 110 111 113
#> [58] 115 116 117 118 122 124 131 132 135 138 140 141 143 144 145 146 148 149


# single prob
prob <- .7
d <- iris
lapply(prob, function(i) {
  sort(sample(1:nrow(d), round(nrow(d) * i), FALSE))
})
#> [[1]]
#>   [1]   1   2   3   6   7   9  10  11  15  16  17  18  20  21  22  23  25  26
#>  [19]  29  30  32  33  34  35  36  37  38  40  43  44  47  48  50  51  53  54
#>  [37]  56  57  59  60  61  62  64  66  67  68  70  71  72  73  74  75  77  79
#>  [55]  80  81  82  83  84  86  87  92  93  94  96  97  98  99 100 101 103 105
#>  [73] 108 109 110 111 112 113 114 115 116 117 118 119 121 122 123 124 125 127
#>  [91] 129 130 131 133 134 136 137 138 139 140 141 143 144 147 149
```

<sup>Created on 2022-06-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Or a bit cleaner:

``` r
d <- iris

prob <- c(.3, .5)
setNames(lapply(prob, function(i) {
  training <- sort(sample(1:nrow(d), round(nrow(d) * i), FALSE))
  test <- setdiff(1:nrow(d), training)
  list(training = training, test = test)
}), sprintf("Probability: %g%%", 100 * prob))
#> $`Probability: 30%`
#> $`Probability: 30%`$training
#>  [1]   3   6   9  11  14  15  21  24  25  26  27  32  34  35  38  39  43  44  45
#> [20]  52  54  58  72  78  81  88  91  95  98 100 101 103 105 117 118 122 123 126
#> [39] 132 134 138 141 142 147 149
#> 
#> $`Probability: 30%`$test
#>   [1]   1   2   4   5   7   8  10  12  13  16  17  18  19  20  22  23  28  29
#>  [19]  30  31  33  36  37  40  41  42  46  47  48  49  50  51  53  55  56  57
#>  [37]  59  60  61  62  63  64  65  66  67  68  69  70  71  73  74  75  76  77
#>  [55]  79  80  82  83  84  85  86  87  89  90  92  93  94  96  97  99 102 104
#>  [73] 106 107 108 109 110 111 112 113 114 115 116 119 120 121 124 125 127 128
#>  [91] 129 130 131 133 135 136 137 139 140 143 144 145 146 148 150
#> 
#> 
#> $`Probability: 50%`
#> $`Probability: 50%`$training
#>  [1]   1   2   4   6   8  10  11  12  14  15  17  18  24  27  28  29  31  32  35
#> [20]  37  41  42  44  46  47  48  50  52  54  58  60  63  64  65  66  67  70  76
#> [39]  81  85  86  88  89  92  94  95  96  97  98  99 100 105 107 108 110 111 112
#> [58] 113 114 115 119 124 130 131 132 134 135 137 138 141 142 143 144 145 147
#> 
#> $`Probability: 50%`$test
#>  [1]   3   5   7   9  13  16  19  20  21  22  23  25  26  30  33  34  36  38  39
#> [20]  40  43  45  49  51  53  55  56  57  59  61  62  68  69  71  72  73  74  75
#> [39]  77  78  79  80  82  83  84  87  90  91  93 101 102 103 104 106 109 116 117
#> [58] 118 120 121 122 123 125 126 127 128 129 133 136 139 140 146 148 149 150


prob <- .7
setNames(lapply(prob, function(i) {
  training <- sort(sample(1:nrow(d), round(nrow(d) * i), FALSE))
  test <- setdiff(1:nrow(d), training)
  list(training = training, test = test)
}), sprintf("Probability: %g%%", 100 * prob))
#> $`Probability: 70%`
#> $`Probability: 70%`$training
#>   [1]   2   3   4   6   7   9  10  12  13  15  16  18  19  20  22  24  25  26
#>  [19]  27  28  29  30  31  32  34  35  36  39  40  41  43  45  46  47  48  49
#>  [37]  50  52  53  56  57  59  62  64  66  67  68  69  70  71  73  75  76  78
#>  [55]  79  80  83  84  86  87  88  89  90  91  92  94  95  98  99 101 102 103
#>  [73] 104 105 106 110 111 112 114 115 118 119 121 122 123 124 125 126 128 129
#>  [91] 130 132 133 134 135 139 140 141 142 143 144 145 146 148 149
#> 
#> $`Probability: 70%`$test
#>  [1]   1   5   8  11  14  17  21  23  33  37  38  42  44  51  54  55  58  60  61
#> [20]  63  65  72  74  77  81  82  85  93  96  97 100 107 108 109 113 116 117 120
#> [39] 127 131 136 137 138 147 150
```

<sup>Created on 2022-06-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
(less confusing code)
``` r
d <- iris

prob <- c(.3, .5)
lapply(prob, function(i) {
  training <- sort(sample(1:nrow(d), round(nrow(d) * i), FALSE))
  test <- setdiff(1:nrow(d), training)
  list(training = training, test = test)
})
#> [[1]]
#> [[1]]$training
#>  [1]   1  13  15  16  20  24  31  37  38  44  52  53  54  56  58  61  62  64  66
#> [20]  70  72  75  77  79  82  89  91  96  98 100 102 106 113 115 120 121 125 126
#> [39] 128 129 130 135 138 141 145
#> 
#> [[1]]$test
#>   [1]   2   3   4   5   6   7   8   9  10  11  12  14  17  18  19  21  22  23
#>  [19]  25  26  27  28  29  30  32  33  34  35  36  39  40  41  42  43  45  46
#>  [37]  47  48  49  50  51  55  57  59  60  63  65  67  68  69  71  73  74  76
#>  [55]  78  80  81  83  84  85  86  87  88  90  92  93  94  95  97  99 101 103
#>  [73] 104 105 107 108 109 110 111 112 114 116 117 118 119 122 123 124 127 131
#>  [91] 132 133 134 136 137 139 140 142 143 144 146 147 148 149 150
#> 
#> 
#> [[2]]
#> [[2]]$training
#>  [1]   2   5  10  12  13  15  16  17  18  19  20  21  24  25  26  27  28  31  36
#> [20]  37  38  43  45  48  50  51  52  57  59  60  61  64  65  66  67  68  72  74
#> [39]  75  79  80  82  86  87  89  90  93  94  96  97  98  99 102 103 104 107 108
#> [58] 109 113 114 115 116 117 118 122 125 131 134 137 139 141 144 145 147 150
#> 
#> [[2]]$test
#>  [1]   1   3   4   6   7   8   9  11  14  22  23  29  30  32  33  34  35  39  40
#> [20]  41  42  44  46  47  49  53  54  55  56  58  62  63  69  70  71  73  76  77
#> [39]  78  81  83  84  85  88  91  92  95 100 101 105 106 110 111 112 119 120 121
#> [58] 123 124 126 127 128 129 130 132 133 135 136 138 140 142 143 146 148 149
```

<sup>Created on 2022-06-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>